File: dayTwo/java/BookSuggestionSystemTest.java
import org.junit.jupiter.api.Test;

import org.junit.jupiter.api.BeforeEach;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.ArrayList;
import java.util.List;

public class BookSuggestionSystemTest {
    BookSuggestionSystem bookSuggestionSystem;

    @BeforeEach
    void setUp() {
        bookSuggestionSystem = new BookSuggestionSystem();
    }

    @Test
    public void testThatBookCountIsInitiallyZero() {
        assertEquals(0, bookSuggestionSystem.getBookCount());
    }

    @Test
    public void testThatAddBookPutsTheBookInTheListAndIncreasesBookCountBy1() {

        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I");
        assertEquals(1, bookSuggestionSystem.getBookCount());

        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book II");
        assertEquals(2, bookSuggestionSystem.getBookCount());

        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book III");
        assertEquals(3, bookSuggestionSystem.getBookCount());
    }

    @Test
    public void testThatAddBookThrowsErrorIfTheBookIsAlreadyInTheList() {
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I");
        assertEquals(1, bookSuggestionSystem.getBookCount());

        assertThrows(
            IllegalArgumentException.class,
            ()-> bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I")
        );
        assertEquals(1, bookSuggestionSystem.getBookCount());
    }

    @Test
    public void testThatRemoveBookRemovesTheBookInTheListAndReducesBookCountBy1() {

        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I");
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book II");
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book III");
        assertEquals(3, bookSuggestionSystem.getBookCount());

        bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book I");
        assertEquals(2, bookSuggestionSystem.getBookCount());

        bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book II");
        assertEquals(1, bookSuggestionSystem.getBookCount());

        bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book III");
        assertEquals(0, bookSuggestionSystem.getBookCount());
    }

    @Test
    public void testThatRemoveBookThrowsErrorIfTheBookIsNotInTheList() {
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I");
        assertEquals(1, bookSuggestionSystem.getBookCount());

        bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book I");
        assertEquals(0, bookSuggestionSystem.getBookCount());

        assertThrows(
            IllegalArgumentException.class,
            ()-> bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book I")
        );
        assertEquals(0, bookSuggestionSystem.getBookCount());
    }

    @Test
    public void testThatGetBookListReturnsListOfAllTheBooks() {

        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I");
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book II");
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book III");

        List<String> expected = List.of(
            "Avatar: The Last Airbender Book I",
            "Avatar: The Last Airbender Book II",
            "Avatar: The Last Airbender Book III"
        );
        List<String> actual = bookSuggestionSystem.getBookList();
        assertEquals(expected, actual);

        bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book I");
        bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book II");
        bookSuggestionSystem.removeBook("Avatar: The Last Airbender Book III");

        expected = new ArrayList<String>();
        actual = bookSuggestionSystem.getBookList();
        assertEquals(expected, actual);
    }

    @Test
    public void testThatUpdateBookUpdateThatParticularBookInTheList() {

        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I");
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book II");

        List<String> expected = List.of(
            "Avatar: The Last Airbender Book I",
            "Avatar: The Last Airbender Book II"
        );
        List<String> actual = bookSuggestionSystem.getBookList();
        assertEquals(expected, actual);

        bookSuggestionSystem.updateBook(
            "Avatar: The Last Airbender Book I",
            "Avatar: The Legend of Korra Book I"
        );
        expected = List.of(
            "Avatar: The Legend of Korra Book I",
            "Avatar: The Last Airbender Book II"
        );
        actual = bookSuggestionSystem.getBookList();
        assertEquals(expected, actual);

        bookSuggestionSystem.updateBook(
            "Avatar: The Last Airbender Book II",
            "Avatar: The Legend of Korra Book II"
        );
        expected = List.of(
            "Avatar: The Legend of Korra Book I",
            "Avatar: The Legend of Korra Book II"
        );
        actual = bookSuggestionSystem.getBookList();
        assertEquals(expected, actual);
    }

    @Test
    public void testThatUpdateBookThrowsErrorIfTheBookIsNotInTheList() {
        bookSuggestionSystem.addBook("Avatar: The Last Airbender Book I");
        bookSuggestionSystem.updateBook(
            "Avatar: The Last Airbender Book I",
            "Avatar: The Legend of Korra Book I"
        );

        assertThrows(
            IllegalArgumentException.class,
            ()-> bookSuggestionSystem.updateBook(
                "Avatar: The Last Airbender Book I",
                "Avatar: The Legend of Korra Book II"
            )
        );
    }
}

File: dayTwo/python/book_suggestion_system_test.py
from unittest import TestCase
from book_suggestion_system import BookSuggestionSystem


class TestBookSuggestionSystem(TestCase):

    def setUp(self):
        self.book_suggestion_system = BookSuggestionSystem()

    def test_book_count_is_initially_zero(self):
        self.assertEqual(0, self.book_suggestion_system.get_book_count())

    def test_I_add_book_it_increases_book_count_by_1(self):

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book I"
        )
        self.assertEqual(1, self.book_suggestion_system.get_book_count())

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book II"
        )
        self.assertEqual(2, self.book_suggestion_system.get_book_count())

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book III"
        )
        self.assertEqual(3, self.book_suggestion_system.get_book_count())

    def test_I_add_book_it_throws_error_if_book_exists(self):

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book I"
        )
        self.assertEqual(1, self.book_suggestion_system.get_book_count())

        self.assertRaises(
            ValueError,
            self.book_suggestion_system.add_book,
            "Avatar: The Last Airbender Book I"
        )
        self.assertEqual(1, self.book_suggestion_system.get_book_count())

    def test_I_remove_book_it_reduces_book_count_by_1(self):

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book I"
        )
        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book II"
        )
        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book III"
        )

        self.assertEqual(3, self.book_suggestion_system.get_book_count())

        self.book_suggestion_system.remove_book(
            "Avatar: The Last Airbender Book I"
        )
        self.assertEqual(2, self.book_suggestion_system.get_book_count())

        self.book_suggestion_system.remove_book(
            "Avatar: The Last Airbender Book II"
        )
        self.assertEqual(1, self.book_suggestion_system.get_book_count())

        self.book_suggestion_system.remove_book(
            "Avatar: The Last Airbender Book III"
        )
        self.assertEqual(0, self.book_suggestion_system.get_book_count())

    def test_I_remove_book_throws_error_if_book_not_found(self):

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book I"
        )
        self.assertEqual(1, self.book_suggestion_system.get_book_count())

        self.book_suggestion_system.remove_book(
            "Avatar: The Last Airbender Book I"
        )
        self.assertEqual(0, self.book_suggestion_system.get_book_count())

        self.assertRaises(
            ValueError,
            self.book_suggestion_system.remove_book,
            "Avatar: The Last Airbender Book I"
        )
        self.assertEqual(0, self.book_suggestion_system.get_book_count())

    def test_get_book_list_returns_all_books(self):

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book I"
        )
        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book II"
        )
        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book III"
        )

        expected = [
            "Avatar: The Last Airbender Book I",
            "Avatar: The Last Airbender Book II",
            "Avatar: The Last Airbender Book III"
        ]

        actual = self.book_suggestion_system.get_book_list()

        self.assertEqual(expected, actual)

        self.book_suggestion_system.remove_book(
            "Avatar: The Last Airbender Book I"
        )
        self.book_suggestion_system.remove_book(
            "Avatar: The Last Airbender Book II"
        )
        self.book_suggestion_system.remove_book(
            "Avatar: The Last Airbender Book III"
        )

        expected = []
        actual = self.book_suggestion_system.get_book_list()

        self.assertEqual(expected, actual)

    def test_I_update_book_it_updates_the_particular_book(self):

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book I"
        )
        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book II"
        )

        expected = [
            "Avatar: The Last Airbender Book I",
            "Avatar: The Last Airbender Book II"
        ]

        actual = self.book_suggestion_system.get_book_list()

        self.assertEqual(expected, actual)

        self.book_suggestion_system.update_book(
            "Avatar: The Last Airbender Book I",
            "Avatar: The Legend of Korra Book I"
        )

        expected = [
            "Avatar: The Legend of Korra Book I",
            "Avatar: The Last Airbender Book II"
        ]

        actual = self.book_suggestion_system.get_book_list()

        self.assertEqual(expected, actual)

        self.book_suggestion_system.update_book(
            "Avatar: The Last Airbender Book II",
            "Avatar: The Legend of Korra Book II"
        )

        expected = [
            "Avatar: The Legend of Korra Book I",
            "Avatar: The Legend of Korra Book II"
        ]

        actual = self.book_suggestion_system.get_book_list()

        self.assertEqual(expected, actual)

    def test_I_update_book_it_throws_error_if_book_not_found(self):

        self.book_suggestion_system.add_book(
            "Avatar: The Last Airbender Book I"
        )

        self.book_suggestion_system.update_book(
            "Avatar: The Last Airbender Book I",
            "Avatar: The Legend of Korra Book I"
        )

        self.assertRaises(
            ValueError,
            self.book_suggestion_system.update_book,
            "Avatar: The Last Airbender Book I",
            "Avatar: The Legend of Korra Book II"
        )
